Migrate all specs to the newer expect syntax

myabc opened the pull request:
Our specs are currently a mix of the older should syntax and the newer expect syntax. I wanted the excuse to give [Transpec](http://yujinakayama.me/transpec/) a try – as you can see, it seems to do a good job. I needed to make very a handful of subsequent fixes.

This relates to https://www.openproject.org/work_packages/4044 – and may assist with the upgrade to RSpec 3.

myabc committed:
Bump RSpec to ~2.14; bundle update

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Convert specs to RSpec 2.14.8 syntax with Transpec

This conversion is done by Transpec 1.10.4 with the following command:
    transpec

* 1414 conversions
    from: obj.should
      to: expect(obj).to

* 646 conversions
    from: == expected
      to: eq(expected)

* 376 conversions
    from: obj.stub(:message)
      to: allow(obj).to receive(:message)

* 107 conversions
    from: obj.should_not
      to: expect(obj).not_to

* 70 conversions
    from: obj.should_receive(:message)
      to: expect(obj).to receive(:message)

* 45 conversions
    from: =~ [1, 2]
      to: match_array([1, 2])

* 27 conversions
    from: lambda { }.should
      to: expect { }.to

* 13 conversions
    from: Klass.any_instance.stub(:message)
      to: allow_any_instance_of(Klass).to receive(:message)

* 8 conversions
    from: === expected
      to: be === expected

* 7 conversions
    from: expect { }.not_to raise_error(SpecificErrorClass)
      to: expect { }.not_to raise_error

* 6 conversions
    from: =~ /pattern/
      to: match(/pattern/)

* 2 conversions
    from: obj.should_not_receive(:message)
      to: expect(obj).not_to receive(:message)

* 1 conversion
    from: < expected
      to: be < expected

* 1 conversion
    from: obj.stub!(:message)
      to: allow(obj).to receive(:message)

* 1 conversion
    from: stub('something')
      to: double('something')

myabc committed:
Fix disable_flash_sweep spec helper

Broken in previous commit 26ade602:

RSpec::Mocks.allow_message is a lower-level API method, allowing us
to define a stub outside of an ExampleGroup context.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use disable_flash_sweep in UsersController spec

Helper fixed in previous commit e31bec17.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix args passed to image_tag in ViewHelpers spec

Deprecated negative raise_error expectations were removed by Transpec
in previous commit 26ade602: this example was previously masking an
ArgumentError.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Mark a failing wiki routing spec as pending

RSpec should syntax masked an issue with the route_to matcher on a
separate line not being called.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Configure RSpec to disable should syntax

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Replace #=~ (operator matcher) with #match_array

RSpec expect syntax no longer supports this operator matcher.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
n.b. This PR doesn't cover cukes, but they could be migrated in a similar fashion.

hschink commented:
@myabc In [`eb70332`](https://github.com/opf/openproject/commit/eb70332a) the route spec uses `post` instead of `get`. I hope that fixes the issue.

myabc committed:
Fix pending wiki routing spec: correct HTTP verb

* remove query params from expectation, since routing specs do not
  test these.

/ HT: @hschink

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Revert bundle update; Bump rspec-rails version

The previous bundle update of all dependencies broke execution of unit
tests: I've yet to determine a more specific cause.

This partially reverts commit 6ea4ed825f231bd3db47e5a73c7c3874f7deb861.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
Issue appears to be with simplecov:

```
bundle update
be ruby -v test/functional/mail_handler_controller_test.rb
```

See PR #1066 for Gemfile fix.

hschink commented:
@myabc I'm sorry but this doesn't work with plug-ins.

myabc commented:
@hschink  OK. Sorry for that. I will revert 4774aeb, so this work can be merged. There is a ticket for converting syntax in plugins: https://www.openproject.org/work_packages/5674

myabc committed:
Revert "Configure RSpec to disable should syntax"

This will break specs on installations with plugins. Reverting until
popular plugins are also migrated to the new RSpec syntax
(see new task: https://www.openproject.org/work_packages/5674)

This reverts commit 4774aeb4ee70b546c37a0a94e565678272bdf042.

hschink commented on Gemfile.lock:
```diff
@@ -265,22 +265,23 @@ GEM
+    rspec-mocks (2.14.6)
+    rspec-rails (2.14.2)
       actionpack (>= 3.0)
+      activemodel (>= 3.0)
```
Do you know where this new dependency comes from?

myabc replied:
@hschink it's now a dependency of `rspec-rails`, but it's been part of our stack since moving to Rails 3, as it's also a required dependency of Rails/Active Record.

See: https://github.com/rspec/rspec-rails/commit/4549965af169019c7f3d679371e4650057d176b2

hschink commented on spec/controllers/api/v2/planning_element_type_colors_controller_spec.rb:
```diff
@@ -80,20 +80,20 @@ def fetch
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
+          expect do
             get 'show', :id => '1337', :format => 'xml'
```
This `get` is equal to the ones above. We may save some line when we extract those `get` statements into a single `before` statement. Additionally, we may get rid of the extraordinary `expect do end` syntax I've never seen before.

myabc replied:
@hschink :+1: that would definitely be a DRY'er option.

`expect` with a block has been in RSpec for a while. According to [this article](http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax#unification_of_block_vs_value_syntaxes), the introduction of the expect syntax was an attempt at unifying block and value syntaxes. I usually use `expect` with braces rather than `do..end` though.

hschink commented on spec/controllers/api/v2/planning_element_types_controller_spec.rb:
```diff
@@ -51,21 +51,21 @@ def fetch
       describe 'with unknown project' do
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
+          expect do
```
Again this extraordinary usage: We should get rid of th because we don't use it elsewhere.

hschink commented on spec/controllers/api/v2/planning_element_types_controller_spec.rb:
```diff
@@ -114,17 +114,17 @@ def fetch
       describe 'with unknown project' do
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
+          expect do
```
As I am not the one who introduced this strange `lambda` syntax, I advocate the punishment of the responsible.

hschink commented on spec/controllers/api/v2/planning_element_types_controller_spec.rb:
```diff
@@ -114,17 +114,17 @@ def fetch
       describe 'with unknown planning element type' do
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
+          expect do
```
:alien:

hschink commented on spec/controllers/api/v2/planning_element_types_controller_spec.rb:
```diff
@@ -134,9 +134,9 @@ def fetch
 
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
+          expect do
```
I should get payed by line... :moneybag:

hschink commented on spec/controllers/api/v2/planning_element_types_controller_spec.rb:
```diff
@@ -207,20 +207,20 @@ def fetch
         else # but have to write it that way
           it 'raises ActiveRecord::RecordNotFound errors' do
-            lambda do
+            expect do
```
I need a :coffee:.

myabc replied:
mit :cake:

hschink commented on spec/controllers/api/v2/planning_elements_controller_spec.rb:
```diff
@@ -526,9 +526,9 @@ def expect_redirect_to
           become_member_with_view_planning_element_permissions
 
           it 'raises ActiveRecord::RecordNotFound errors' do
-            lambda do
```
Give me a :cocktail: please.

hschink commented on spec/controllers/api/v2/planning_elements_controller_spec.rb:
```diff
@@ -786,9 +786,9 @@ def expect_no_content
           become_member_with_delete_planning_element_permissions
 
           it 'raises ActiveRecord::RecordNotFound errors' do
-            lambda do
```
Are you kidding me :cat2: ?

hschink commented on spec/controllers/api/v2/planning_elements_controller_spec.rb:
```diff
@@ -823,20 +823,20 @@ def expect_no_content
 
           it 'deletes the record' do
             get 'destroy', :project_id => project.id, :id => planning_element.id, :format => 'xml'
-            lambda do
```
I hope this is the last one...

hschink commented on spec/controllers/api/v2/project_associations_controller_spec.rb:
```diff
@@ -124,9 +124,9 @@ def fetch
 
         describe 'w/ the current user being a member' do
           it 'raises ActiveRecord::RecordNotFound errors' do
-            lambda do
```
...no, it isn't. :gun:

hschink commented on spec/controllers/api/v2/project_types_controller_spec.rb:
```diff
@@ -77,9 +77,9 @@ def fetch
   describe 'show.xml' do
     describe 'with unknown project type' do
       it 'raises ActiveRecord::RecordNotFound errors' do
-        lambda do
```
:alien:

hschink commented on spec/controllers/api/v2/projects_controller_spec.rb:
```diff
@@ -106,9 +106,9 @@ def fetch
 
       describe 'with unknown project' do
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
:alien:

hschink commented on spec/controllers/api/v2/reported_project_statuses_controller_spec.rb:
```diff
@@ -46,9 +46,9 @@
     describe 'index.xml' do
       describe 'with unknown project_type' do
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
:alien:

hschink commented on spec/controllers/api/v2/reported_project_statuses_controller_spec.rb:
```diff
@@ -106,17 +106,17 @@
       describe 'with unknown project_type' do
 
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
If I get the one responsible in my hands...

hschink commented on spec/controllers/api/v2/reported_project_statuses_controller_spec.rb:
```diff
@@ -106,17 +106,17 @@
 
       describe 'with unknown reported_project_status' do
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
:beer:

hschink commented on spec/controllers/api/v2/reported_project_statuses_controller_spec.rb:
```diff
@@ -133,18 +133,18 @@
         end
 
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
:rage1:

hschink commented on spec/controllers/api/v2/reported_project_statuses_controller_spec.rb:
```diff
@@ -133,18 +133,18 @@
       describe 'with reported_project_status not available for project_type' do
         it 'raises ActiveRecord::RecordNotFound errors' do
           available_reported_project_status
-          lambda do
```
:rage2:

hschink commented on spec/controllers/api/v2/reported_project_statuses_controller_spec.rb:
```diff
@@ -207,9 +207,9 @@
     describe 'show.xml' do
       describe 'with unknown reported_project_status' do
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
It's not even the half of this PR...

hschink commented on spec/controllers/api/v2/reported_project_statuses_controller_spec.rb:
```diff
@@ -221,9 +221,9 @@
         end
 
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
Who has reviewed this (and accepted) this?!

hschink commented on spec/controllers/api/v2/reportings_controller_spec.rb:
```diff
@@ -131,9 +131,9 @@ def fetch
 
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
+          expect do
```
:rage3:

hschink commented on spec/controllers/api/v2/timelines_controller_spec.rb:
```diff
@@ -165,9 +165,9 @@ def self.become_member_with_delete_permissions
           become_member_with_all_permissions
 
           it 'raises ActiveRecord::RecordNotFound errors' do
-            lambda do
```
:rage4:

hschink commented on spec/controllers/api/v2/timelines_controller_spec.rb:
```diff
@@ -189,9 +189,9 @@ def self.become_member_with_delete_permissions
         let(:other_project)  { FactoryGirl.create(:project, :identifier => 'other') }
 
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
Oh man...

hschink commented on spec/controllers/timelines_controller_spec.rb:
```diff
@@ -165,9 +165,9 @@ def self.become_member_with_delete_permissions
           become_member_with_all_permissions
 
           it 'raises ActiveRecord::RecordNotFound errors' do
-            lambda do
```
:hammer:

hschink commented on spec/controllers/timelines_controller_spec.rb:
```diff
@@ -189,9 +189,9 @@ def self.become_member_with_delete_permissions
         let(:other_project)  { FactoryGirl.create(:project, :identifier => 'other') }
 
         it 'raises ActiveRecord::RecordNotFound errors' do
-          lambda do
```
:rabbit2: just for you @myabc

hschink commented on spec/controllers/work_packages/bulk_controller_spec.rb:
```diff
@@ -262,7 +263,7 @@
           subject { WorkPackage.find_all_by_priority_id(priority.id).collect(&:id) }
 
-          it { should =~ work_package_ids }
+          it { should match_array(work_package_ids) }
```
I just want to be sure: In this case `should` is still allowed?

myabc replied:
@hschink Yes. In RSpec 2.14 you can still use should with the One-liner syntax with `should`. RSpec 3 introduces `is_expected` as an alternative though: https://www.relishapp.com/rspec/rspec-core/v/3-0/docs/subject/one-liner-syntax

hschink commented on spec/models/mail_handler_spec.rb:
```diff
@@ -283,27 +283,27 @@
     Setting.default_language = 'en'
     Role.non_member.update_attribute :permissions, [:add_work_packages]
     project.update_attribute :is_public, true
-    lambda do
```
Wow, what a mess... :running:

myabc replied:
See 9335752ef80bc23bcf027fae1e92b5b5400b4094

hschink commented on spec/models/user_spec.rb:
```diff
@@ -243,25 +243,25 @@ def build_user_double_with_expired_password(is_expired)
       end
 
       it 'creates a SystemUser' do
-        lambda do
```
:feelsgood:

hschink commented on spec/models/user_spec.rb:
```diff
@@ -243,25 +243,25 @@ def build_user_double_with_expired_password(is_expired)
       end
 
       it 'returns existing SystemUser'  do
-        lambda do
```
:eyes:

myabc committed:
Attempt to work-around an AccountController CI failure

     NoMethodError:
       undefined method `original_path_set' for nil:NilClass

Failure only occurs with a full plugin configuration.

See https://github.com/rspec/rspec-rails/issues/860

hschink commented:
@myabc it's still failing on internal CI: This time `allow_password_changes?` is called on `nil`.

myabc commented:
@hschink That's good. It means it's now reproducing what I'm seeing locally. I'll take another look.

myabc committed:
Fix failing AccountController#register spec

With an OpenProject config containing private plugins this spec will
fail with the following Error:

    ActionView::Template::Error: undefined method `allow_password_changes?' for nil:NilClass
    ./app/models/user.rb:377:in `change_password_allowed?'

This failure was most likely previously obscured because
`render_views` was not enabled.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@hschink OK fix posted.

myabc committed:
Merge branch 'dev' into fix/rspec-expect-syntax

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Conflicts:
	spec/models/permitted_params_spec.rb
	spec/models/work_package/work_package_acts_as_journalized_spec.rb
	spec/models/work_package/work_package_planning_spec.rb

myabc committed:
Convert specs to RSpec 2.14.8 syntax with Transpec

This conversion is done by Transpec 1.10.4 with the following command:
    transpec spec/models/permitted_params_spec.rb

* 10 conversions
    from: obj.should
      to: expect(obj).to

* 10 conversions
    from: == expected
      to: eq(expected)

* 1 conversion
    from: obj.should_receive(:message)
      to: expect(obj).to receive(:message)
